src/embind/emval.js

```diff
@@ -12,39 +12,43 @@
 /*jslint sub:true*/ /* The symbols 'fromWireType' and 'toWireType' must be accessed via array notation to be closure-safe since craftInvokerFunction crafts functions as strings that can't be closured. */
 
 // -- jshint doesn't understand library syntax, so we need to mark the symbols exposed here
-/*global getStringOrSymbol, emval_handles, Emval, __emval_unregister, count_emval_handles, emval_symbols, __emval_decref*/
+/*global getStringOrSymbol, emval_freelist, emval_handles, Emval, __emval_unregister, count_emval_handles, emval_symbols, __emval_decref*/
 /*global emval_addMethodCaller, emval_methodCallers, addToLibrary, global, emval_lookupTypes, makeLegalFunctionName*/
 /*global emval_get_global*/
 
+// Number of handles reserved for non-use (0) or common values w/o refcount.
+{{{ 
+  globalThis.EMVAL_RESERVED_HANDLES = 5;
+  globalThis.EMVAL_LAST_RESERVED_HANDLE = globalThis.EMVAL_RESERVED_HANDLES - 1;
+  null;
+}}}
 var LibraryEmVal = {
-  $emval_handles__deps: ['$HandleAllocator'],
-  $emval_handles: "new HandleAllocator();",
+  // Stack of handles available for reuse.
+  $emval_freelist: [],
+  // Array of alternating pairs (value, refcount).
+  $emval_handles: [],
   $emval_symbols: {}, // address -> string
 
   $init_emval__deps: ['$count_emval_handles', '$emval_handles'],
   $init_emval__postset: 'init_emval();',
   $init_emval: () => {
-    // reserve some special values. These never get de-allocated.
-    // The HandleAllocator takes care of reserving zero.
-    emval_handles.allocated.push(
-      {value: undefined},
-      {value: null},
-      {value: true},
-      {value: false},
+    // reserve 0 and some special values. These never get de-allocated.
+    emval_handles.push(
+      0, 1,
+      undefined, 1,
+      null, 1,
+      true, 1,
+      false, 1,
     );
-    Object.assign(emval_handles, /** @lends {emval_handles} */ { reserved: emval_handles.allocated.length }),
+  #if ASSERTIONS
+    assert(emval_handles.length / 2 === {{{ EMVAL_RESERVED_HANDLES }}});
```

Comment:
I think it should be possible to remove all the `/ 2` and `* 2` business for further slightly smaller code size and micro-perf win. The fact that the returned handles would "only" have a 1 billion range instead of 2 billion range is likely not ever possible to exhaust - that 1 billion range would already mean more than 16 gigabytes of memory usage in the emval_handles array. (2*double for each handle).

Works either way though.

Reply:
Oh I see, you mean handing back only even numbers as a handles.   Nice idea!

(Then I guess we can assert that handles are even when we receive them from the user).

Reply:
Updated as suggested, did require updates to the .h file for the new even handles for reserved values.